lib/elixir/lib/module/types.ex

```diff
@@ -1,87 +1,284 @@
+
+            context =
+              update_in(context.local_sigs, &Map.put(&1, fun_arity, {kind, inferred, mapping}))
+
+            {kind, inferred, context}
+
+          false ->
+            false
         end
+    end
+  end
+
+  defp local_handler({fun, arity} = fun_arity, kind, meta, clauses, mode, stack, context) do
+    expected = List.duplicate(Descr.dynamic(), arity)
+    stack = stack |> fresh_stack(mode, fun_arity) |> with_file_meta(meta)
+
+    {_, _, mapping, clauses_types, clauses_context} =
+      Enum.reduce(clauses, {0, 0, [], [], context}, fn
+        {meta, args, guards, body}, {index, total, mapping, inferred, context} ->
+          context = fresh_context(context)
+
+          try do
+            {args_types, context} =
+              Pattern.of_head(args, guards, expected, :default, meta, stack, context)
+
+            {return_type, context} =
+              Expr.of_expr(body, stack, context)
+
+            {type_index, inferred} =
+              add_inferred(inferred, args_types, return_type, total - 1, [])
+
+            if type_index == -1 do
+              {index + 1, total + 1, [{index, total} | mapping], inferred, context}
+            else
+              {index + 1, total, [{index, type_index} | mapping], inferred, context}
+            end
+          rescue
+            e ->
+              internal_error!(e, __STACKTRACE__, kind, meta, fun, args, guards, body, stack)
+          end
       end)
-    end)
+
+    inferred = {:infer, Enum.reverse(clauses_types)}
+    {inferred, mapping, restore_context(context, clauses_context)}
   end
 
-  defp with_file_meta(meta, file) do
+  # We check for term equality of types as an optimization
+  # to reduce the amount of check we do at runtime.
+  defp add_inferred([{args, existing_return} | tail], args, return, index, acc),
+    do: {index, Enum.reverse(acc, [{args, Descr.union(existing_return, return)} | tail])}
+
+  defp add_inferred([head | tail], args, return, index, acc),
+    do: add_inferred(tail, args, return, index - 1, [head | acc])
+
+  defp add_inferred([], args, return, -1, acc),
+    do: {-1, [{args, return} | Enum.reverse(acc)]}
+
+  defp with_file_meta(stack, meta) do
     case Keyword.fetch(meta, :file) do
-      {:ok, {meta_file, _}} -> meta_file
-      :error -> file
+      {:ok, {meta_file, _}} -> %{stack | file: meta_file}
+      :error -> stack
     end
   end
 
-  defp internal_error!(e, stack, kind, meta, module, fun, args, guards, body) do
+  defp internal_error!(e, trace, kind, meta, fun, args, guards, body, stack) do
     def_expr = {kind, meta, [guards_to_expr(guards, {fun, [], args}), [do: body]]}
 
     exception =
       RuntimeError.exception("""
-      found error while checking types for #{Exception.format_mfa(module, fun, length(args))}:
+      found error while checking types for #{Exception.format_mfa(stack.module, fun, length(args))}:
 
       #{Exception.format_banner(:error, e, stack)}\
```

Comment:
I believe this should be `Exception.format_banner(:error, e, trace)` ? Dialyzer is warning about that:

```
The call 'Elixir.Exception':format_banner
         ('error',
          _e@1 ::
              #{'__exception__' := 'true',
                '__struct__' := atom(),
                atom() => _},
          _stack@1 ::
              #{'function' := {_, _},
                'mode' := 'dynamic' | 'infer' | 'traversal',
                _ => _}) breaks the contract 
          (kind(), any(), stacktrace()) -> 'Elixir.String':t()
```

Reply:
Thank you for the report!

I tried to prove that we could catch this error too. Check this out. If I do this diff:

```diff
--- a/lib/elixir/lib/exception.ex
+++ b/lib/elixir/lib/exception.ex
@@ -131,7 +131,7 @@ def normalize(_kind, payload, _stacktrace), do: payload
   @spec format_banner(kind, any, stacktrace) :: String.t()
   def format_banner(kind, exception, stacktrace \\ [])

-  def format_banner(:error, exception, stacktrace) do
+  def format_banner(:error, exception, [_ | _] = stacktrace) do
     exception = normalize(:error, exception, stacktrace)
     "** (" <> inspect(exception.__struct__) <> ") " <> message(exception)
   end
diff --git a/lib/elixir/lib/module/types.ex b/lib/elixir/lib/module/types.ex
index acfb2f7d6..6976e6579 100644
--- a/lib/elixir/lib/module/types.ex
+++ b/lib/elixir/lib/module/types.ex
@@ -277,7 +277,7 @@ defp with_file_meta(stack, meta) do
     end
   end

-  defp internal_error!(e, trace, kind, meta, fun, args, guards, body, stack) do
+  defp internal_error!(e, trace, kind, meta, fun, args, guards, body, %{} = stack) do
     def_expr = {kind, meta, [guards_to_expr(guards, {fun, [], args}), [do: body]]}

     exception =
```

I get this report:

```

     warning: incompatible types given to Exception.format_banner/3:

         Exception.format_banner(:error, e, stack)

     given types:

         :error, dynamic(), dynamic(%{...})

     but expected one of:

         #1
         dynamic(:error), dynamic(), dynamic(non_empty_list(term(), term()))

         #2
         dynamic(:throw), dynamic(), dynamic()

         #3
         dynamic(:exit), dynamic(), dynamic()

         #4
         dynamic({:EXIT, term()}), dynamic(), dynamic()

     where "e" was given the type:

         # type: dynamic()
         # from: lib/elixir/lib/module/types.ex:280:24
         e

     where "stack" was given the type:

         # type: dynamic(%{...})
         # from: lib/elixir/lib/module/types.ex:280:75
         %{} = stack

     typing violation found at:
     │
 287 │       #{Exception.format_banner(:error, e, stack)}\
     │                   ~
     │
     └─ (elixir 1.18.0-dev) lib/elixir/lib/module/types.ex:287:19: Module.Types.internal_error!/9
```

Which is nice!

I think however Dialyzer is typechecking across the function boundaries, because that's the only way it could know that stack has mode `:dynamic | :infer | :traversal`. I will play with some of the ideas later. :)